[Wesley Yu] iP

headcube1 opened the pull request:
# HeadCube
> "Chatbots will be your new best friend.” — Christine Crandell ([Forbes](https://www.forbes.com/sites/christinecrandell/2016/10/23/chatbots-will-be-your-new-best-friend/?sh=70565620a245))

HeadCube helps you keep tracks of the things you need to do. It's

- text-based
- simple to use
- _EXTREMELY FAST_

All you need to do is,

1. download it
2. add your things
3. let it manage :grinning:

The application is **FREE**!

The features include :
- [x] Managing tasks
- [x] Deadlines
- [ ] Reminders

## 
If you can program in Java, you can practice! Here is the `main` method:
``` java
public class HeadCube {
    public static void main(String[] args) {
        Application.launch(HeadCube.class, args);
    }
}
  ```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

headcube1 committed:
Changed name as well as added greeting and exit

headcube1 committed:
Added Echo method

headcube1 committed:
Added add and list

headcube1 committed:
Created Task class as well as mark as done feature

headcube1 committed:
Todos, events, deadlines

headcube1 committed:
Dealt with 2 types of errors

headcube1 committed:
added delete method

headcube1 committed:
Text-ui-test

headcube1 committed:
Add Save and Load methods

headcube1 committed:
Add Dates and Times

headcube1 committed:
Merge branch 'master' into branch-Level-8

headcube1 committed:
Merge branch 'branch-Level-8'

headcube1 committed:
Add more use of OOP

headcube1 committed:
Add headcube package

headcube1 committed:
Merge branch 'add-gradle-support'

headcube1 committed:
Add JUnit Tests

headcube1 committed:
Add JavaDocs

headcube1 committed:
Add find method

headcube1 committed:
Merge branch 'branch-Level-9'

headcube1 committed:
Use CheckStyle

headcube1 committed:
Merge branch 'branch-A-CheckStyle'

headcube1 committed:
Add GUI

headcube1 committed:
Add GUI

headcube1 committed:
Merge branch 'branch-Level-10'

Dethada commented on src/main/java/headcube/Deadlines.java:
```diff
@@ -0,0 +1,62 @@
+    public Deadlines(String description, String by) {
+        super(description);
+        DateTimeFormatter targetFormatter = DateTimeFormatter.ofPattern("MMM dd yyyy");
+        DateTimeFormatter originalFormatter = DateTimeFormatter.ofPattern("yyyy-MM-dd");
```
Consider naming `originalFormatter` to `alternativeFormatter` instead since it will only be used if `targetFormatter` fails

Dethada commented on src/main/java/headcube/Deadlines.java:
```diff
@@ -0,0 +1,62 @@
+     */
+    public Deadlines(String description, String by) {
+        super(description);
+        DateTimeFormatter targetFormatter = DateTimeFormatter.ofPattern("MMM dd yyyy");
```
Consider making `targetFormatter` a private field to allow reuse, instead of having to recreate the object multiple times below

Dethada commented on src/main/java/headcube/Events.java:
```diff
@@ -0,0 +1,66 @@
+     */
+    public Events(String description, String start, String end) {
+        super(description);
+        DateTimeFormatter targetFormatter = DateTimeFormatter.ofPattern("MMM dd yyyy HH:mm");
```
Similar to the `targetFormatter` in `Deadlines.java`, consider making this a private field

Dethada commented on src/main/java/headcube/Events.java:
```diff
@@ -0,0 +1,66 @@
+    public Events(String description, String start, String end) {
+        super(description);
+        DateTimeFormatter targetFormatter = DateTimeFormatter.ofPattern("MMM dd yyyy HH:mm");
+        DateTimeFormatter originalFormatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm");
```
Similar to the `originalFormatter` in `Deadlines.java`, consider renaming to `alternativeFormatter`

Dethada commented on src/main/java/headcube/MainWindow.java:
```diff
@@ -0,0 +1,67 @@
+    private void handleUserInput() {
+        String input = userInput.getText();
+
+        if (input.equals("bye")) {
+            dialogContainer.getChildren().addAll(
+                    DialogBox.getUserDialog(input, userImage),
+                    DialogBox.getHeadCubeDialog(Ui.exit(), headcubeImage)
+            );
+            userInput.clear();
+        } else {
+            String response = headcube.getParser().parse(input);
+            dialogContainer.getChildren().addAll(
+                    DialogBox.getUserDialog(input, userImage),
+                    DialogBox.getHeadCubeDialog(response, headcubeImage)
+            );
+            userInput.clear();
+        }
```
The body in the 2 branches of the conditional is almost the same, consider extracting them out to make the code more concise, here's an idea
```java
String response = input.equals("bye") ? Ui.exit() : headcube.getParser().parse(input);
dialogContainer.getChildren().addAll(
        DialogBox.getUserDialog(input, userImage),
        DialogBox.getHeadCubeDialog(response, headcubeImage)
);
userInput.clear();
```

Dethada commented on src/main/java/headcube/Parser.java:
```diff
@@ -0,0 +1,86 @@
+     */
+    public String parse(String input) {
+        try {
+            String[] split = input.split(" ", 2);
```
Since `split` is a rather generic name, I think better name would be `commandAndArgs`.
Also consider to store the result to variables like `command` and `args` instead of repeatedly accessing `split[0]` and `split[1]`

Dethada commented on src/main/java/headcube/Parser.java:
```diff
@@ -0,0 +1,86 @@
+     * @param input The user input to parse.
+     * @return The response to the input
+     */
+    public String parse(String input) {
```
`parse` might not be the best name for this method since it is also generating the response to the user in addition to parsing the input.

Dethada commented on src/main/java/headcube/Parser.java:
```diff
@@ -0,0 +1,86 @@
+            if (input.equals("list")) {
+                return ui.list(taskList);
+            } else if (split[0].equals("mark")) {
+                return taskList.mark(Integer.parseInt(split[1]));
```
Consider handling [NumberFormatException](https://docs.oracle.com/javase/8/docs/api/java/lang/NumberFormatException.html) - if the string does not contain a parsable integer. There are other cases of this below too.

Dethada replied:
Also consider the case where the integer is not a valid task index, e.g. -1 or 9999999

Dethada commented on src/main/java/headcube/Parser.java:
```diff
@@ -0,0 +1,86 @@
+                }
+
+                StringBuilder sb = new StringBuilder();
+                sb.append("Got it. I've added this task:\n  ")
+                        .append(taskList.get(taskList.size() - 1))
+                        .append("\nNow you have ")
+                        .append(taskList.size())
+                        .append(" tasks in the list.\n");
```
I like the formatting here

Dethada commented on src/main/java/headcube/Storage.java:
```diff
@@ -0,0 +1,116 @@
+     */
+    public String save(TaskList taskList) {
+        try {
+            String directoryPath = "./data";
+            String filePath = directoryPath + "/HeadCube.txt";
```
Consider storing these as fields, would be helpful to allow access to the filePath in the `load` method below

Dethada commented on src/main/java/headcube/Storage.java:
```diff
@@ -0,0 +1,116 @@
+
+            FileWriter fw = new FileWriter(file);
+            for (Task task : taskList.getList()) {
+                fw.write(task.toFileFormat() + System.lineSeparator());
```
I like the use of S`ystem.lineSeparator()` for cross platform compatibility.

Dethada commented on src/main/java/headcube/Storage.java:
```diff
@@ -0,0 +1,116 @@
+
+        Task task = null;
+
+        if ("T".equals(event)) {
```
Consider storing `"T"` to a constant instead of hardcoding it here

Dethada commented on src/main/java/headcube/Storage.java:
```diff
@@ -0,0 +1,116 @@
+
+        if ("T".equals(event)) {
+            task = new ToDos(description);
+        } else if ("D".equals(event)) {
+            if (parts.length > 3) {
+                String by = parts[3].replace("(by: ", "").replace(")", "");
+                task = new Deadlines(description, by);
+            }
```
can be merged to `else if ("D".equals(event) && (parts.length > 3))`

Dethada commented on src/main/java/headcube/Storage.java:
```diff
@@ -0,0 +1,116 @@
+                String by = parts[3].replace("(by: ", "").replace(")", "");
+                task = new Deadlines(description, by);
+            }
+        } else {
+            if (parts.length > 3) {
+                String timeInfo = parts[3].replace("(from: ", "").replace(")", "");
+                String[] times = timeInfo.split(" to: ");
+                String start = times[0];
+                String end = times.length > 1 ? times[1] : "";
+                task = new Events(description, start, end);
+            }
```
can be merged to `else if (parts.length > 3)`

Dethada reviewed (commented):
Looks good overall, other than some minor issues. I like the structure of the code, would be even better if packages are also used for organization.

garywongkai commented on src/main/java/headcube/Ui.java:
```diff
@@ -0,0 +1,85 @@
+        }
+        return sb.toString();
+    }
+}
```
Method names like list and showFoundTasks are descriptive,
Maybe consider using verbs that describe actions for method names. For example, list could be listTasks, and showFoundTasks could be showMatchingTasks, but just minor changes👍

garywongkai commented on src/main/java/headcube/Task.java:
```diff
@@ -0,0 +1,59 @@
+    public String toString() {
+        return "[" + getStatusIcon() + "] " + description;
+    }
+}
```
Great job, can see that all of your naming conventions are following camelCase 👍

garywongkai commented on src/main/java/headcube/Deadlines.java:
```diff
@@ -0,0 +1,62 @@
+     * @param description The description of the deadline task.
+     * @param by The deadline date in "MMM dd yyyy" or "yyyy-MM-dd" format.
+     */
+    public Deadlines(String description, String by) {
```
Can consider renaming "by" to be "by_date" for better clarify :)

garywongkai reviewed (commented):
Great job 💯  , nothing much to comment about aside from some minor readability changes and naming conventions :D

headcube1 committed:
Add Assertions

There were certain assumptions to be included at various
points of the code. Hence, adding assertions was needed.

By adding assertions, it can be verfied at runtime which can
indicate a bug.

Assertions were placed at crucial points of the code.

headcube1 committed:
Add Assertions

There were certain assumptions to be included at various
points of the code. Hence, adding assertions was needed.

By adding assertions, it can be verfied at runtime which can
indicate a bug.

Assertions were placed at crucial points of the code.

headcube1 committed:
Improve Code Quality

The need to improve code quality is crucial in ensuring
run-time efficiency, security and readability.

Remove Magic numbers as well as making the code more obvious to
ensure the quality is higher.

headcube1 committed:
Merge pull request #2 from headcube1/branch-A-CodeQuality

Improve Code Quality

headcube1 committed:
Merge branch 'master' into branch-A-Assertions

headcube1 committed:
Merge pull request #1 from headcube1/branch-A-Assertions

Add Assertions

headcube1 committed:
no message

headcube1 committed:
Add duplicate dectection

The ability to detect same tasks input as well as if the input is
already in the tasklist. Deny the user of adding duplicate tasks.

This is to ensure the user is not spamming the tasklist with
duplicates.

headcube1 committed:
Merge branch 'branch-C-DetectDuplicates'

headcube1 committed:
Test MD

headcube1 committed:
Add UG

headcube1 committed:
no message

headcube1 committed:
no message
